firestore-translate-text/__tests__/functions.test.ts

```diff
@@ -0,0 +1,384 @@
+      mockTranslateClass.mockClear();
+
+      require("../functions/src");
+
+      expect(mockTranslateClass).toHaveBeenCalledTimes(1);
+      expect(mockTranslateClass).toHaveBeenCalledWith({
+        projectId: defaultEnvironment.PROJECT_ID,
+      });
+    });
+
+    test("function skips deleted document change events", async () => {
+      mockFirestoreUpdate.mockClear();
+      mockTranslateClassMethod.mockClear();
+
+      documentChange.after.exists = false;
+      const callResult = await wrappedFunction(documentChange);
+
+      expect(callResult).toBeUndefined();
+      expect(mockConsoleLog).toHaveBeenCalledWith(
+        "Document was deleted, no processing is required"
+      );
+
+      expect(mockTranslateClassMethod).not.toHaveBeenCalled();
+      expect(mockFirestoreUpdate).not.toHaveBeenCalled();
+    });
+
+    test("function skips 'update' document change events if the input is unchanged", async () => {
+      mockFirestoreUpdate.mockClear();
+      mockTranslateClassMethod.mockClear();
+
+      beforeSnapshot = functionsTest.firestore.makeDocumentSnapshot(
+        { input: "hello" },
+        "translations/id1"
+      );
+
+      afterSnapshot = functionsTest.firestore.makeDocumentSnapshot(
+        { input: "hello", changed: 123 },
+        "translations/id1"
+      );
+
+      documentChange = functionsTest.makeChange(beforeSnapshot, afterSnapshot);
+
+      const callResult = await wrappedFunction(documentChange);
+      expect(callResult).toBeUndefined();
+      expect(mockConsoleLog).toHaveBeenCalledWith(
+        "Document was updated, input string has not changed, no processing is required"
+      );
+
+      expect(mockTranslateClassMethod).not.toHaveBeenCalled();
+      expect(mockFirestoreUpdate).not.toHaveBeenCalled();
+    });
+
+    test("function skips document 'created' document change events without any input", async () => {
+      mockFirestoreUpdate.mockClear();
+      mockTranslateClassMethod.mockClear();
+
+      afterSnapshot = functionsTest.firestore.makeDocumentSnapshot(
+        { changed: 123 },
+        "translations/id1"
+      );
+      documentChange = functionsTest.makeChange(beforeSnapshot, afterSnapshot);
+
+      const callResult = await wrappedFunction(documentChange);
+
+      expect(callResult).toBeUndefined();
+
+      expect(mockConsoleLog).toHaveBeenCalledWith(
+        "Document was created without an input string, no processing is required"
+      );
+
+      expect(mockTranslateClassMethod).not.toHaveBeenCalled();
+      expect(mockFirestoreUpdate).not.toHaveBeenCalled();
+    });
+
+    test("function exits early if input & output fields are the same", async () => {
+      jest.resetModules();
+
+      restoreEnv();
+      restoreEnv = mockedEnv({
+        ...defaultEnvironment,
+        INPUT_FIELD_NAME: "input",
+        OUTPUT_FIELD_NAME: "input",
+      });
+
+      wrappedFunction = functionsTest.wrap(
+        require("../functions/src").fstranslate
+      );
```

Comment:
I haven't used jest before so perhaps this is an dumb question but this logic seems like its already being handled in the `beforeEach` calls earlier and I was confused as to why some of it is being repeated.  Why was this necessary?

Reply:
As it is currently implemented,`beforeEach` runs generic environment setup logic. This is fine for most tests but in this instance, we wanted `INPUT_FIELD_NAME ` & `OUTPUT_FIELD_NAME ` to have the same value. So we set the environment variables again. If you see logic that looks duplicated, it's a safe bet that the test requires custom setup outside of the generic environment setup.

Reply:
This makes sense. I think this is worth a comment or putting in a function w/ a descriptive name like `ensureInputOutputFieldsAreTheSame()` or something more concise.